core/runtime/common/trie_storage_provider_impl.cpp

```diff
@@ -60,10 +74,40 @@ namespace kagome::runtime {
   }
 
   outcome::result<common::Buffer> TrieStorageProviderImpl::forceCommit() {
-    if(persistent_batch_ != nullptr) {
+    if (persistent_batch_ != nullptr) {
       return persistent_batch_->commit();
     }
     return common::Buffer{};
   }
 
+  outcome::result<void> TrieStorageProviderImpl::startTransaction() {
+    stack_of_batches_.emplace(current_batch_);
+    current_batch_ =
+        std::make_shared<TopperTrieBatchImpl>(std::move(current_batch_));
+    return outcome::success();
+  }
+
+  outcome::result<void> TrieStorageProviderImpl::rollbackTransaction() {
+    if (stack_of_batches_.empty()) {
+      return TransactionError::NO_TRANSACTIONS_WERE_STARTED;
+    }
+
+    current_batch_ = std::move(stack_of_batches_.top());
+    return outcome::success();
+  }
+
+  outcome::result<void> TrieStorageProviderImpl::commitTransaction() {
+    if (stack_of_batches_.empty()) {
+      return TransactionError::NO_TRANSACTIONS_WERE_STARTED;
+    }
+
+    auto commitee_batch =
+        std::dynamic_pointer_cast<TopperTrieBatch>(current_batch_);
+    BOOST_ASSERT(commitee_batch != nullptr);
+    OUTCOME_TRY(commitee_batch->writeBack());
+
+    current_batch_ = std::move(stack_of_batches_.top());
```

Comment:
Shouldn't the top of the stacked be popped here?

Reply:
Sure! Thanks for your notice